test/unit/util/locking/RedisResourceLocker.test.ts

```diff
@@ -0,0 +1,194 @@
+/* eslint-disable jest/valid-expect-in-promise */
+// eslint-disable-next-line import/default
+import redis from 'redis';
+import type { RedisClient } from 'redis';
+import { Lock } from 'redlock';
+import { InternalServerError } from '../../../../src/util/errors/InternalServerError';
+import { RedisResourceLocker } from '../../../../src/util/locking/RedisResourceLocker';
+
+describe('A RedisResourceLocker', (): void => {
+  let locker: any;
+  const identifier = { path: 'http://test.com/foo' };
+  beforeEach(async(): Promise<void> => {
+    // Mocking redis this way because of this issue: https://github.com/facebook/jest/issues/8983
+    // eslint-disable-next-line import/no-named-as-default-member
+    redis.createClient = jest.fn().mockReturnValue({
+      on: jest.fn(),
+      quit: jest.fn(),
+      get: jest.fn(),
+    });
+
+    locker = new RedisResourceLocker([ '6379' ]);
+    locker.redlock.lock = jest.fn().mockImplementation(
```

Comment:
The issue above is about mocking `redis` but mocking `redlock` should still work right? Then you shouldn't need to cast it to `any` so you can overwrite a private field.

Reply:
I don't fully understand what you are saying.
By adding `jest.mock('redlock')` to the file, I am still not able to use this `locker.redlock.lock = jest.fn()`.
Or do you want me to mock it like `redlock.lock = jest.fn()`?

Reply:
You can see 4 different ways on how to do this here: https://jestjs.io/docs/es6-class-mocks#manual-mock

I'm guessing it would look something like this instead:
```typescript
jest.mock('redlock', (): any => jest.fn().mockImplementation((): any => ({
  lock: jest.fn().mockImplementation(async(resource: string, ttl: number): Promise<Lock> =>
    new Lock({}, resource, 'test', ttl, 100)),
  unlock: jest.fn(),
  extend: jest.fn().mockImplementation(
    async(lock: Lock, ttl: number): Promise<Lock> => {
      lock.expiration += ttl;
      return lock;
    },
  ),
})));
```
With probably a change in what gets passed to the `new Lock` call (I don't know what it needs access to, or you could mock Lock as well).

A smaller example in the code base already can be found at https://github.com/solid/community-server/blob/main/test/unit/init/CliRunner.test.ts#L17

Reply:
I am not sure what [that example](https://github.com/solid/community-server/blob/main/test/unit/init/CliRunner.test.ts#L17) has in common with this issue.
I tried mocking it your way but that didn't fix anything as it is still just mocking. The issue (is it really?) is that typescript checks typings 'n stuff and complains that the field is private.

I did some googling: many stackoverflows suggest using a cast to any as the code will be compiled to js anyways (right?).
The way I originally did it was never recommended tho (`let locker: any;`). They suggested (in this scenario) replacing all calls like `locker.redlock.lock` to `(locker as any).redlock.lock`. 
I tried this and it seems to work.

I would love to see it work without any `any` casts, but I think my knowledge of jest mocking is too limited.

Since this is the last open conversation, would you mind fixing it? This way I can clearly see how it should be done, as it is based on code I am familiar with.

Reply:
Will have a look later, was also already thinking that it might be best if I look into this since it's the only remaining part and the mocking can get fiddly in these cases.